dhall/ghc-src/Dhall/Import/UserHeaders.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE LambdaCase           #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE NamedFieldPuns       #-}
+
+module Dhall.Import.UserHeaders
+    ( defaultUserHeaders
+    , envOnlyUserHeaders
+    ) where
+
+import Control.Exception        (tryJust)
+import Control.Monad            (guard)
+import Data.Either.Combinators  (rightToMaybe)
+import Data.Text                (Text)
+import Dhall.Import.Headers     (SiteHeadersFile(..))
+import Dhall.Core
+    ( Directory(..)
+    , File(..)
+    , FilePrefix(..)
+    , ImportType
+    )
+
+import System.Directory         (getXdgDirectory, XdgDirectory(XdgConfig))
+import System.Environment       (lookupEnv)
+import System.FilePath          ((</>), splitDirectories)
+import System.IO.Error          (isDoesNotExistError)
+
+import qualified Data.Text      as Text
+import qualified Data.Text.IO   as IO
+import qualified Dhall.Core     as Core
+
+siteHeadersFile :: FilePath -> ImportType -> Text -> SiteHeadersFile
+siteHeadersFile parentDirectory source fileContents =
+    SiteHeadersFile { parentDirectory , source, fileContents }
+
+-- lift 'siteHeadersFile' to work on IO (Maybe Text)
+siteHeadersFile' :: FilePath -> ImportType -> IO (Maybe Text) -> IO (Maybe SiteHeadersFile)
```

Comment:
I believe if you replace all occurrences of `IO (Maybe a)` with `MaybeT IO a` it will simplify several things you're doing here

For example, you would no longer need the `siteHeadersFile'` function because it's (basically) just `(>>=)` for `MaybeT`.  That means that `envOnlyUserHeaders` would simplify to:

```haskell
envOnlyUserHeaders :: MaybeT IO SiteHeadersFile
envOnlyUserHeaders = do
    string <- liftIO (lookupEnv key)

    siteHeadersFile "." (Core.Env (Text.pack key)) (Text.pack string)
  where
    key = "DHALL_HEADERS"
```

Similarly, `configFileOnlyUserHeaders` would simplify to:

```haskell
configFileOnlyUserHeaders = do
    directory <- liftIO (getXdgDirectory XdgConfig "dhall")

    text <- liftIO (tryReadFile (directory </> (Text.unpack filename)))

    siteHeadersFile directory (makeSource directory) text
```

… and `defaultUserHeaders` would simplify to:

```haskell
defaultUserHeaders = envOnlyUserHeaders <|> configOnlyUserHeaders
```